File: gc-starter/log-spring-boot-starter/src/main/java/com/gc/starter/log/aspect/LogAspect.java
package com.gc.starter.log.aspect;

import com.baomidou.mybatisplus.core.toolkit.IdWorker;
import com.gc.common.base.exception.handler.GlobalExceptionHandler;
import com.gc.common.base.message.Result;
import com.gc.common.base.utils.AopUtils;
import com.gc.common.base.utils.IpUtils;
import com.gc.common.base.utils.JsonUtils;
import com.gc.starter.log.LogProperties;
import com.gc.starter.log.annotation.Log;
import com.gc.starter.log.constants.LogIdentConstant;
import com.gc.starter.log.handler.LogHandler;
import com.gc.starter.log.model.SysLogPO;
import com.google.common.collect.Lists;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.BooleanUtils;
import org.apache.commons.lang3.ObjectUtils;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.Signature;
import org.aspectj.lang.annotation.*;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.lang.Nullable;
import org.springframework.util.StringUtils;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

import javax.annotation.PostConstruct;
import javax.servlet.http.HttpServletRequest;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * 日志切面
 * @author shizhongming
 * 2020/1/22 1:54 下午
 */
@Aspect
@Slf4j
public final class LogAspect {

    /**
     * 保存日志的编码集合
     */
    private List<Integer> codeList;

    @Autowired
    private LogProperties logProperties;

    @Autowired
    private LogHandler logHandler;

    @PostConstruct
    public void init() {
        if (this.codeList == null) {
            String codes = this.logProperties.getCodes();
            if (StringUtils.isEmpty(codes)) {
                this.codeList = Lists.newArrayList();
            } else {
                this.codeList = Arrays.stream(codes.split(","))
                        .map(code -> Integer.parseInt(code.trim()))
                        .collect(Collectors.toList());
            }
        }
    }

    /**
     * 切面
     */
    @Pointcut("@annotation(com.gc.starter.log.annotation.Log)")
    public void logPointCut() {
        // 切点
    }

    /**
     * 环绕
     * @param point 切入点
     * @return 结果
     */
    @Around("logPointCut()")
    public Object around(ProceedingJoinPoint point) throws Throwable {
        long beginTime = System.currentTimeMillis();
        Exception exception = null;
        Object result = null;
        try {
            // 执行方法
            result = point.proceed();
        } catch (Exception e) {
            exception = e;
        }
        // 用时（毫秒）
        long time = System.currentTimeMillis() - beginTime;
        if (BooleanUtils.isTrue(this.logProperties.getConsole())) {
            // 执行耗时
            log.info("Time-Consuming : {} ms", time);
        }
        if (ObjectUtils.isNotEmpty(exception)) {
            result = GlobalExceptionHandler.doException(exception);
        }
        this.saveLog(point, time, result);
        // 保存日志
        return result;
    }

    /**
     * 执行前
     * @param point 切点
     */
    @Before("logPointCut()")
    private void before(JoinPoint point) {
        if (BooleanUtils.isTrue(this.logProperties.getConsole())) {
            ServletRequestAttributes attributes = (ServletRequestAttributes) RequestContextHolder.getRequestAttributes();
            HttpServletRequest request = Objects.requireNonNull(attributes).getRequest();
            log.info("========================================== Start ==========================================");
            // 打印请求 url
            log.info("URL            : {}", request.getRequestURL().toString());
            // 打印 Http method
            log.info("HTTP Method    : {}", request.getMethod());
            // 打印调用 controller 的全路径以及执行方法
            log.info("Class Method   : {}.{}", point.getSignature().getDeclaringTypeName(), point.getSignature().getName());
            // 打印请求的 IP
            log.info("IP             : {}", IpUtils.getIpAddr());
        }
    }

    @After("logPointCut()")
    private void after() {
        if (BooleanUtils.isTrue(this.logProperties.getConsole())) {
            log.info("=========================================== End ===========================================");
        }
    }

    /**
     * 保存日志
     * @param point 切入点
     * @param time 用时
     * @param result 执行结果
     */
    @SuppressWarnings("rawtypes")
    private void saveLog(ProceedingJoinPoint point, long time, Object result) {
        try {
            final Signature signature = point.getSignature();
            if (signature instanceof MethodSignature) {
                final Method method = ((MethodSignature) signature).getMethod();
                final Log logAnnotation = method.getAnnotation(Log.class);
                int code = 200;
                boolean saveLog = true;
                // 错误信息
                String errorMessage = null;
                if (result instanceof Result) {
                    code = ((Result) result).getCode();
                    if (BooleanUtils.isFalse(((Result) result).getOk())) {
                        errorMessage = ((Result) result).getMessage();
                    }
                    List<Integer> saveCodeList = this.getCodeList();
                    // 如果设置了保存的编码，并且不包含保存的编码  则不保存日志
                    if (!saveCodeList.isEmpty() && !saveCodeList.contains(code)) {
                        saveLog = false;
                    }
                }
                if (saveLog) {
                    this.doSaveLog(point, result, signature, logAnnotation, time, code, errorMessage);
                }

            }
        } catch (Exception e) {
            log.error("保存日志发生错误", e);
        }
    }

    /**
     * 执行保存日志
     * @param point 切面
     * @param signature signature
     * @param logAnnotation 日志注解
     * @param time 用户
     * @param code 编码
     * @param errorMessage 错误信息
     */
    private void doSaveLog(ProceedingJoinPoint point, @Nullable Object result, Signature signature, Log logAnnotation, long time, int code, String errorMessage) {
        // 请求的方法名
        final String className = point.getTarget().getClass().getName();
        final String methodName = signature.getName();
        // 设置请求参数
        String parameter = JsonUtils.toJsonString(AopUtils.getParameterMap(point));
        final SysLogPO sysLog = SysLogPO.builder()
                .logId(IdWorker.getId())
                .operation(logAnnotation.value())
                .useTime(time)
                .method(String.join(".", className, methodName))
                .ip(IpUtils.getIpAddr())
                .params(parameter)
                .requestPath(((ServletRequestAttributes) Objects.requireNonNull(RequestContextHolder.getRequestAttributes())).getRequest().getServletPath())
                .statusCode(code)
                .errorMessage(errorMessage)
                .type(logAnnotation.type().name())
                .ident(LogIdentConstant.AUTO.getValue())
                // todo：待处理
                .platform(null)
                .build();
        if (logAnnotation.saveResult() && Objects.nonNull(result)) {
            sysLog.setResult(JsonUtils.toJsonString(result));
        }
        this.logHandler.save(sysLog, point, logAnnotation, time, code, result, errorMessage);
    }

    /**
     * 获取编码集合
     * @return 允许的编码集合
     */
    private List<Integer> getCodeList() {
        return this.codeList;
    }
}

File: smart-starter/log-spring-boot-starter/src/main/kotlin/com/smart/starter/log/aspect/LogAspect.kt
package com.smart.starter.log.aspect

import com.alibaba.fastjson.JSON
import com.smart.auth.common.utils.AuthUtils
import com.smart.common.log.annotation.Log
import com.smart.common.message.Result
import com.smart.common.utils.IPUtils
import com.smart.starter.log.model.SysLogDO
import com.smart.starter.log.service.SysLogService
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.aspectj.lang.reflect.MethodSignature
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.context.request.ServletRequestAttributes
import java.util.*

/**
 * 日志切面
 * @author ming
 * TODO:异常捕获
 * 2019/6/28 下午4:23
 */
@Aspect
class LogAspect {

    @Autowired
    private lateinit var logService: SysLogService

    /**
     * 切面
     */
    @Pointcut("@annotation(com.smart.common.log.annotation.Log)")
    fun logPointCut() {
    }

    @Around("logPointCut()")
    @Throws(Throwable::class)
    fun around(point: ProceedingJoinPoint): Any? {
        val beginTime = System.currentTimeMillis()
        // 执行方法
        val result = point.proceed()
        // 执行时长(毫秒)
        val time = System.currentTimeMillis() - beginTime
        //异步保存日志
        this.saveLog(point, time, result)
        return result
    }

    @Throws(InterruptedException::class)
    private fun saveLog(joinPoint: ProceedingJoinPoint, time: Long, result: Any) {
        val sysLog = SysLogDO()
        // 设置用户
        sysLog.userId = AuthUtils.getCurrentUserId()
        // 获取注解日志描述
        val signature = joinPoint.signature as MethodSignature
        val method = signature.method
        val syslog = method.getAnnotation(Log::class.java)
        if (syslog != null) {
            // 注解上的描述
            sysLog.operation = syslog.value
        }
        // 设置用时
        sysLog.useTime = time.toInt()
        // 请求的方法名
        val className = joinPoint.target.javaClass.name
        val methodName = signature.name
        sysLog.method = "$className.$methodName()"
        // 请求参数
        val args = joinPoint.args
        if (args.isNotEmpty()) {
            sysLog.params = JSON.toJSONString(args[0])
        }
        // 获取IP地址
        val request = (RequestContextHolder.getRequestAttributes() as ServletRequestAttributes).request
        sysLog.ip = IPUtils.getIpAddr(request)
        sysLog.createTime = Date()
        sysLog.requestPath = request.servletPath
        // 设置code码
        if (result is Result<*>) {
            result.code?.let {
                sysLog.statusCode = it
            }
            if (!result.ok) {
                sysLog.errorMessage = result.message
            }
        }
        this.logService.save(sysLog)
    }
}